HPCC-12395 Redis plugin

jamienoss opened the pull request:
See README.

Signed-off-by: james james.noss@lexisnexis.com

jamienoss committed:
HPCC-12395 Redis plugin

See README.

Signed-off-by: james <james.noss@lexisnexis.com>

richardkchapman commented on cmake_modules/commonSetup.cmake:
```diff
@@ -76,13 +76,14 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
 
   option(USE_MYSQL "Enable MySQL support" ON)
-  option(USE_CASSANDRA "Enable Cassandra support" ON)
+  option(USE_CASSANDRA "Enable Cassandra support" OFF)
```
You should not release PR with this disabled.

jamienoss replied:
Very sorry, I had stated this in the commit message. DON'T merge this PR please.

richardkchapman replied:
I won't

richardkchapman commented on esp/src/dgrid:
```diff
@@ -1 +1 @@
-Subproject commit 24564450a03aa08cc98c2ac89d77d99d7c1be8c3
+Subproject commit 62b4967eb2c8e3dd8c52a02ce73093989cb8c7ca
```
These do not belong in this commit

richardkchapman commented on plugins/cassandra/cpp-driver:
```diff
@@ -1 +1 @@
-Subproject commit 189e46e3386cc2be0fd7428c1072339b9c611d6e
+Subproject commit 5822a469fdca3bdc0ca034dd61c01d1966acc6ef
```
Nor does this

richardkchapman commented on plugins/redis/README:
```diff
@@ -0,0 +1,122 @@
+/**********************/
+/*    Reds Plugin     */
```
Typo

richardkchapman commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,859 @@
+
+#include "hiredis/async.h"
+
+namespace Lock
```
You should not be using namespaces to differentiate parts of the code (and then reusing names within those namespaces) - it's VERY confusing!

All the code in the redis plugin should be in one namespace, and you should use different classnames for the different types.

jamienoss committed:
wip

Signed-off-by: James Noss <james.noss@lexisnexis.com>

jamienoss committed:
Added timeouts to libev event loops.

Signed-off-by: James Noss <james.noss@lexisnexis.com>

jamienoss committed:
Removed multiple namespaces

Signed-off-by: James Noss <james.noss@lexisnexis.com>

jamienoss committed:
Rename keyPtr to lockObject

Signed-off-by: James Noss <james.noss@lexisnexis.com>

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+/**********************/
+(WIP)
+apt-get install redis-server #includes redis-cli
+apt-get insatll libhiredis-dev
```
typo install

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+STRING SetString(UNSIGNED8 keyPtr, CONST STRING value, UNSIGNED4 expire = 0)
+STRING GetString(UNSIGNED8 keyPtr)
+
+Note that SET also returns a value. This is the vaue just set and not a wrapper for the redis command
```
typo value

ghalliday commented on plugins/redis/CMakeLists.txt:
```diff
@@ -0,0 +1,70 @@
+        )
+
+    ADD_DEFINITIONS( -D_USRDLL -DECL_REDIS_EXPORTS)
+    #SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}  -std=gnu++0x")
```
Remove?

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+/**********************/
```
Could rename this README.rst so github applies pretty printing

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+It contains three modules - 'sync', 'async', and 'locking'. Each one containing the core GET/SET variations
+(by ECL type) differing at their base communication layer with the redis servers. 'async' and 'locking' both
+use the asynchronous hiredis API, and 'sync' the synchronous API. Due to the nature of ECL as a declarative
+and non-functional language, no async activity breaches directly into the ECL layer in a usable manner, i.e.
```
I struggled to follow these sentences.

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+handled internally by the redis command 'SELECT DB'.
+
+The core SET functions have the prototype pattern (e.g.)
+SetString(CONST VARSTRING options, CONST VARSTRING key, CONST STRING value, UNSIGNED database = 0,  UNSIGNED4 expire = 0 /*days*/)
```
I would remove CONST probably confuses things (and shouldn't really be needed on the prototype.)

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+'GETSET'. See the section on locks.
+
+Within the sync and async modules, SETs overwrite their key values. The respective GETs (other than for
+STRINGs) will fail upon reuqest for a non-existent key. If you're not sure use EXISTS first (certain
```
request.

ghalliday replied:
@richardkchapman Do we want that semantic of failing if the key doesn't exist?

richardkchapman replied:
I think so. It's the same as memcached?

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+Within the sync and async modules, SETs overwrite their key values. The respective GETs (other than for
+STRINGs) will fail upon reuqest for a non-existent key. If you're not sure use EXISTS first (certain
+precaution has to be taken when using multiple plugin calls within ECL, regarding their actual execution
+timing. Remember ECL is declarative and may decide to hoist your calls.) GetString will return an empty
```
Why is GetString different?

jamienoss replied:
If you meant "how" - they return a empty string (though I'm not sure if I have implemented this correctly for ECL STRINGS).

"Why?"
1) They are the only types that this can apply to.
2) It really helps with the breach of the lock/race logic into ECL.
3) Shamser et al would very much like this (they only intend to use redis with strings).

Is this allowed?

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+ECL clients/threads wish to read the same key, at the same time, from the redis server and encounter an
+non-existent key (cache miss). It is thus desirable to be able to then SET this missing key however, only
+once and by one client/thread. This is such as to prevent redundant evaluation/retrieval of the new value
+within this race condition. The core of this design is to have one instant (the first/race-winner), 'lock'
```
instance

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+Since redis does not implicitly support locks, it is this 'channel', itself, that represents the notion of a
+lock within this plugin library. These channels/locks are a particular string prefixed by the string
+'redis_ecl_lock'. When locked the key's value contains this channel/lock string. This is so lock acknowledging
+GETs can subscribe to the key via the correct channel, being the  locked keys vaue itself.
```
value.  Double space.

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+/**********************/
+/*     Locking        */
+/**********************/
+The locking module was designed with a particular use in mind. This concerns the scenario where multiple
```
This section could do with revisiting.  In some places the text could be simplified.  Also good to making sure it is clear from the point of view of someone who wants to use the system, and doesn't particularly care how it is implemented.

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+
+Based on the above described scenario, ideally this would be implemented by the following function design:
+
+GET(options, key, theECLFunctionToCallUponMiss(params));
```
Is this really helpful?  "We should have implemented it like this but we didn't"

jamienoss replied:
Well... yes & no. This initial  version had you guys as its target audience (only).

That aside, breaching the logic into the ECL doesn't really make sense and is/looks messy.
I wanted to explain this.

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+
+Thus, to use the locking module two options are possible:
+
+UNSIGNED lock := locking.GetLockObject(options, key, database);
```
I may find out later, but I don't currently see where this lock is released.

jamienoss replied:
Yes, this is something that we need to discuss. As I've said before this isn't actually a lock. No redis key is locked within this function. This is used to pass the required info of server:port, db, & key into the locking routines in a single 'lump' so that you can't missmash them as parameters. It then has as it's destructor a call to delete the key if still locked. I believe the plan was to return it as a row, of which kind I'm not sure.

ghalliday commented on plugins/redis/README:
```diff
@@ -0,0 +1,124 @@
+Thus, to use the locking module two options are possible:
+
+UNSIGNED lock := locking.GetLockObject(options, key, database);
+STRING x := locking.GetString(lockObject);
```
do you mean lock?

jamienoss replied:
Yes, cheers.

ghalliday commented on testing/regress/ecl/redislockingtest.ecl:
```diff
@@ -0,0 +1,63 @@
+NOFOLD(MissThenLock(lock));
+
+
+REAL pi := 3.14159265359;
```
This test isn't testing the locking in the way that the user would expect to use it.  It needs to match the real-world example which was included in the README.

jamienoss replied:
Yes, as I mentioned on flowdock, this isn't a test at all. This is the main part that I need help with regarding hoisting. The examples in the README file work so long as that is the _only_ redis plugin functions present due to hoisting. It has no key file. It is all commented out other than some unrelated and the two MissThenLock calls. Monitoring this from redis-cli showed that GetLockObject was being called twice rather than once.

I'll move it to the ecl regress dir for now.

ghalliday commented on plugins/redis/lib_redis.ecllib:
```diff
@@ -0,0 +1,87 @@
+  INTEGER DBSize(CONST VARSTRING options, UNSIGNED database = 0) : cpp,action,context,namespace='RedisPlugin',entrypoint='RDBSize';
+END;
+
+export async := SERVICE : plugin('redis')
```
can you add ,namespace('RedisPlugin') to the service and remove from each function?

jamienoss replied:
I can, very nice.

ghalliday commented on plugins/redis/lib_redis.ecllib:
```diff
@@ -0,0 +1,87 @@
+  DATA GetData(CONST VARSTRING options, CONST VARSTRING key, UNSIGNED database = 0) : cpp,once,context,namespace='RedisPlugin',entrypoint='AsyncRGetData';
+END;
+
+export locking := SERVICE : plugin('redis')
```
Should consider a typedef along the lines of

RedisLock := UNSIGNED;

ghalliday commented on plugins/redis/redisasync.hpp:
```diff
@@ -0,0 +1,65 @@
+
+extern "C++"
+{
+    //------------------------ASYNC--GET----------------------------------------
```
Shouldn't these be inside a namespace?

jamienoss replied:
I guess not, but will do so.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    key.set(_key);
+    channel.set(_channel);
+    database =_database;
+    RedisPlugin::parseOptions(ctx, _options, master, port);
```
It wasn't immediately clear that master/port were set up by this function.  Not sure how you would make it clearer.
One possibility would be to have an RedisOptions class containing master and port and call options.parse(ctx, _options).  Ok as it is though.

jamienoss replied:
I'll do this at it will solve the re-parsing issue also.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    const char * channel2 = reply->query()->str;
+
+    if (strcmp(channel, channel2) == 0)
+        OwnedReply reply = RedisPlugin::createReply(redisCommand(context, "DEL %b", key.str(), strlen(key.str())));
```
Since this is slightly unusual I would enclose in {}

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    redisContext * context = redisConnectWithTimeout(master, port, REDIS_TIMEOUT);
+    CriticalBlock block(crit);
+    OwnedReply reply = RedisPlugin::createReply(redisCommand(context, "GET %b", key.str(), strlen(key.str())));
+    const char * channel2 = reply->query()->str;
```
Could use a better variable name

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+
+class ReturnValue : public CInterface
+{
+public :
```
You could probably use a MemoryAttr to simplify some of this code.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    StringAttr channel;
+};
+
+class SubscriptionThread : implements IThreaded, implements IInterface, public SubContainer
```
Minor: Should it contain a SubContainer, rather than being derived from one?

jamienoss replied:
Originally I wasn't sure I would want to reuse an instance of this class with another callback/redis-operator. I tried to keep a lot of this code (the plugin in general) generic and open as it is far from complete, regarding the incorporation of a lot of the really useful redis commands. Pipelining commands, sets, lists, and allowing for operations server-side such as sorting.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    virtual ~SubscriptionThread()
+    {
+        stopEvLoop();
+        wait(TIMEOUT);//give stopEvLoop time to complete
```
The comment doesn't clarify why the wait is needed.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+
+    inline void setCB(redisCallbackFn * _callback) { callback = _callback; }
+    inline redisCallbackFn * getCB() { return callback; }
+    inline void setChannel(const char * _channel) { channel.set(_channel); }
```
Having lots of get/set functions (or public members) suggests the code isn't encapsulated correctly.

However in this case I think quite a few of these functions aren't used, so should probably be deleted.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    CThreaded  thread;
+    struct ev_loop * evLoop;
+};
+SubContainer::SubContainer(ICodeContext * ctx, const char * options, const char * _channel, unsigned __int64 _database) : AsyncConnection(ctx, options, _database)
```
Why not use the same constructor as below?

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    if (database == 0)
+        return;
+    attachLibev();
+    VStringBuffer cmd("SELECT %llu", database);
```
I know this almost certainly doesn't work in windows, but we tend to avoid using ll and use " I64F " instead

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+}
+void AsyncConnection::createAndAssertConnection(ICodeContext * ctx)
+{
+    context = NULL;
```
Should really be in the constructor initialiser list.

jamienoss committed:
(1) review changes

Signed-off-by: James Noss <james.noss@lexisnexis.com>

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+            VStringBuffer msg("Redis Plugin : failed to create connection context for %s:%d - %s", connection->getMaster(), connection->getPort(), context->errstr);
+            rtlFail(0, msg.str());
+        }
+        {
```
missing else?   (even though it doesn't change what the code does.)

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+        {
+            const AsyncConnection * connection = (AsyncConnection *)context->data;
+            VStringBuffer msg("Redis Plugin : failed to connect to %s:%d - %s", connection->master.str(), connection->port, context->errstr);
+            rtlFail(0, msg.str());
```
Do you want failures in the callback?  Will they get fed back to the correct place?

jamienoss replied:
If the correct place is back within the thread calling ev_loop/ev_run, then yes.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+        assertConnection();
+        //There should always be a connection error
+        VStringBuffer msg("Redis Plugin: %s%s", _msg, "no 'reply' nor connection error");
+        rtlFail(0, msg.str());
```
Should be using an error code other than 0 for all new errors that we report (and going back and adding sensible codes for all previous ones)

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    ev_timer timer;
+    ev_timer_init(&timer, timeoutCB, timeout, 0.);
+    ev_timer_again(evLoop, &timer);
+    ev_run(evLoop, 0);
```
I don't really understand this code.  Shouldn't there be a single thread running the ev event loop?  How does the event loop ever terminate?

jamienoss replied:
ev_run is the actual libev event loop. This is terminated by the timer (setup above) upon a timeout or when there are no more events open. In the case of this plugin such events are the hiredis libev adapter io read and write routines. These are closed within the relevant callbacks thus terminating the loop.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+{
+    attachLibev();
+    assertRedisErr(redisAsyncCommand(context, subCB, (void*)&value, "SUBSCRIBE %b", channel, strlen(channel)), "SUBSCRIBE buffer write error");
+    handleLoop(EV_DEFAULT_ EV_TIMEOUT);
```
I think this needs more explanation.  The code I would expect to see would be

ensureEvLoopRunning()
asyncCommand()
response.wait()

where the callback either signals a semaphore, or interrupts it if there was an error.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+}
+//-------------------------------------------GET-----------------------------------------
+//---OUTER---
+template<class type> void AsyncRGet(ICodeContext * ctx, const char * options, const char * key, type & returnValue, const char * channel, unsigned __int64 database)
```
What are the async versions meant to do?  Is there any advantage to the ECL user?  Should we consider an async function which evaluates early, but is then queried to extract the results.  (Child queries are heading this way with lazy evaluation.)

jamienoss replied:
At the ECL functional level not a tremendous amount as these plugin functions don't breach any asynchronous operations through to the ECL and are thus identical to the sync versions. They do however, allow for performance tinkering of the backend polling operations. The sync commands use hiredis code that has only a while(read) loop that continuously (system)reads the socket. There are no interval setting etc.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    returnSize = retVal.getSize();
+    returnValue = reinterpret_cast<type*>(retVal.getStr());
+}
+void AsyncConnection::getVoidPtrLenPair(ICodeContext * ctx, const char * key, size_t & returnSize, void * & returnValue, const char * channel )
```
Why is this function needed/different from the one above?

jamienoss replied:
I'll look back into it as I can't quite remember. I had compiler/linker issues when using the templates with void*, he wrong function was being called. This was either _me_ having issues or it's an actual issue. I special cased to see what was going on, it worked, I left it.

jamienoss replied:
BTW this is also used within the memcached plugin code.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+{
+    KeyLock * lockObject = (KeyLock*)_lockObject;
+    const char * channel = lockObject->getChannel();
+    if (!lockObject || strlen(channel) == 0)
```
This check for NULL is a bit late.

jamienoss replied:
yep, good spot, cheers.

jamienoss replied:
Yep, I can see that originally strlen(channel) was lockObject->getChannel(). I then moved it out to use else where, without changing the null test, furthermore removing the second use altogether.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+ECL_REDIS_API bool ECL_REDIS_CALL RMissThenLock(ICodeContext * ctx, unsigned __int64 _lockObject)
+{
+    KeyLock * lockObject = (KeyLock*)_lockObject;
+    const char * channel = lockObject->getChannel();
```
Most of this should be a member function of lockObject

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+        rtlFail(0, msg.str());
+    }
+    const char * options = lockObject->getOptions();
+    Owned<AsyncConnection> master = AsyncConnection::createConnection(ctx, options, lockObject->getDatabase());
```
Minor: Shouldn't really need to reparse the options.

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+    KeyLock * lockObject = (KeyLock*)_lockObject;
+    AsyncRSet(ctx, lockObject->getOptions(), lockObject->getKey(), valueLength, value, expire, lockObject->getChannel(), lockObject->getDatabase());
+    returnLength = valueLength;
+    memcpy(&returnValue, value, returnLength);
```
I don't think returnValue hasn't been allocated - this wil crash/overwrite memory

ghalliday replied:
Same an several other functions.

jamienoss replied:
returnValue is a reference to the ECL return values. i.e. STRING. These are already allocated before the primary plugin function call.

jamienoss replied:
A _very_ good point!

ghalliday commented on plugins/redis/redisasync.cpp:
```diff
@@ -0,0 +1,870 @@
+ECL_REDIS_API void ECL_REDIS_CALL LockingRSetStr(ICodeContext * ctx, size32_t & returnLength, char * & returnValue, unsigned __int64 _lockObject, size32_t valueLength, const char * value, unsigned expire)
+{
+    KeyLock * lockObject = (KeyLock*)_lockObject;
+    AsyncRSet(ctx, lockObject->getOptions(), lockObject->getKey(), valueLength, value, expire, lockObject->getChannel(), lockObject->getDatabase());
```
It may not be worth it since it is so small, but the number of lockObject->... suggests this should be a member of KeyLock - if nothing else it is breaking encapsulation.

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+    StringBuffer stats("Redis Plugin : Server stats - ");
+    stats.newline().append(reply->query()->str).newline();
+    ctx->logString(stats.str());
+    alreadyInitialized = true;
```
This variable doesn't appear to be used.  Should it be?

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+    //database is switched only on connection creation allowing for the following condition to redis default
+    if (database == 0)
+        return;
+    VStringBuffer cmd("SELECT %llu", database);
```
See comment on ll and I64F above

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+    OwnedReply reply = createReply(redisCommand(context, "DEL %b %u", key, strlen(key), _expire));
+    assertOnError(reply->query(), "'Expire' request failed - ");
+}
+bool SyncConnection::exist(ICodeContext * ctx, const char * key)
```
picky: Should really be exists

ghalliday replied:
(and all the functions that call it)

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+//--INNER--
+template<class type> void SyncConnection::get(ICodeContext * ctx, const char * key, type & returnValue)
+{
+    OwnedReply reply = createReply(redisCommand(context, "GET %b", key, strlen(key)));
```
Is there an expiry timeout on get, or will this wait forever?

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+}
+ECL_REDIS_API void ECL_REDIS_CALL SyncRGetStr(ICodeContext * ctx, size32_t & returnLength, char * & returnValue, const char * options, const char * key, unsigned __int64 database)
+{
+    size_t _returnLength;
```
returnSize for consistency with the rest

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+}
+ECL_REDIS_API void ECL_REDIS_CALL SyncRGetData(ICodeContext * ctx, size32_t & returnLength, void * & returnValue, const char * options, const char * key, unsigned __int64 database)
+{
+    size_t _returnLength;
```
same here

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+
+SyncConnection::SyncConnection(ICodeContext * ctx, const char * _options, unsigned __int64 _database) : Connection(ctx, _options, _database)
+{
+   context = redisConnectWithTimeout(master.str(), port, REDIS_TIMEOUT);
```
trivial: indented by 3 instead of 4

ghalliday commented on plugins/redis/lib_redis.ecllib:
```diff
@@ -0,0 +1,87 @@
+  DATA GetData(CONST VARSTRING options, CONST VARSTRING key, UNSIGNED database = 0) : cpp,once,context,namespace='RedisPlugin',entrypoint='SyncRGetData';
+
+  BOOLEAN Exists(CONST VARSTRING options, CONST VARSTRING key, UNSIGNED database = 0) : cpp,once,context,namespace='RedisPlugin',entrypoint='RExist';
+  FlushDB(CONST VARSTRING options, UNSIGNED database = 0) : cpp,once,context,namespace='RedisPlugin',entrypoint='RClear';
```
Does flush delete the entire database?  If so a name like RemoveDatabase might be more descriptive.

ghalliday commented on plugins/redis/redissync.cpp:
```diff
@@ -0,0 +1,351 @@
+    assertOnError(reply->query(), appendIfKeyNotFoundMsg(reply->query(), key, keyMsg));
+
+    returnLength = reply->query()->len;
+    size_t returnSize = returnLength*sizeof(type);
```
This doesn't look right.  Why the * sizeof(type)?

ghalliday commented:
@jamienoss I've had a first look.   My impressions are that the sync code is close to ready, I'm not sure what the async code provides for the ECL programmer, and the locked code needs more work.

For the locked code you really need examples that match the intended use.  At a minimum I would suggest:
- using both forms of the expected usage pattern.  
- a test using the patterns to access multiple keys
- a test (using SEQUENTIAL) which accesses the same key more than once
- a test which calls this inside a child query (e.g., using a key value from the input record).

The way that the plugin uses the ev event loop library may be correct, but it isn't what I would expect, so if it is correct it could do with some more explanation of the process.

(For the future, it might have been better to push, review and merge the synchronous code before starting the lock code since it may have caught some things a bit earlier and would be less of a lump of code to try and review in one go.)

jamienoss committed:
(wip) Created RedisServer class to store ip & port helping with
re-parsing of options and link counting instances.

Started to refactor KeyLock...

Compiles & works

Signed-off-by: James Noss <james.noss@lexisnexis.com>
